### ISSUE TITLE
HELLO_REQUEST epoch 0

### ISSUE THREAD

**Issue by boaks** [eclipse/tinydtls#93]
According `hs_attempt_with_existing_peer` a DTLS_HT_HELLO_REQUEST is processed for epoch 0 (mean unencrypted, maybe spoofed). Is that really wanted/intended?

If so, I would prefer a "compile-time switch", because in my opinion such a function makes DTLS vulnerable.

**Comment by boaks** [comment]
[RFC5246, 7.4.1.1.  Hello Request](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.1.1)

> When this message will be sent:
> 
>       The HelloRequest message MAY be sent by the server at any time.
> 
>    Meaning of this message:
> 
>       HelloRequest is a simple notification that the client should begin
>       the negotiation process anew.  In response, the client should send
>       a ClientHello message when convenient.  This message is not
>       intended to establish which side is the client or server but
>       merely to initiate a new negotiation.  Servers SHOULD NOT send a
>       HelloRequest immediately upon the client's initial connection.  It
>       is the client's job to send a ClientHello at that time.

The code in dtls.c

```
if (expected_epoch != msg_epoch) {
  if (hs_attempt_with_existing_peer(msg, rlen, peer)) {
    state = DTLS_STATE_WAIT_CLIENTHELLO;
    role = DTLS_SERVER;
  } else {
    dtls_warn("Wrong epoch, expected %i, got: %i\n",
		    expected_epoch, msg_epoch);
    break;
  }
}
```

makes me belief, that the client send it. 
AFAIK, that's out of RFC5246/RFC6347.

  


```

**Comment by boaks** [comment]
```
case DTLS_HT_HELLO_REQUEST:

    if (state != DTLS_STATE_CONNECTED) {
      /* we should just ignore such packets when in handshake */
      return 0;
    }

    if (peer && !peer->handshake_params) {
      peer->handshake_params = dtls_handshake_new();
      if (!peer->handshake_params)
        return dtls_alert_fatal_create(DTLS_ALERT_INTERNAL_ERROR);

      peer->handshake_params->hs_state.mseq_r = 0;
      peer->handshake_params->hs_state.mseq_s = 0;
    }

    /* send ClientHello with empty Cookie */
    err = dtls_send_client_hello(ctx, peer, NULL, 0);
    if (err < 0) {
      dtls_warn("cannot send ClientHello\n");
      return err;
    }
    peer->state = DTLS_STATE_CLIENTHELLO;
    break;
```

OK, 

```
    state = DTLS_STATE_WAIT_CLIENTHELLO;
    role = DTLS_SERVER;
```

is only used for CLIENT_HELLO. Hard to follow.

**Comment by boaks** [comment]
I'm still confused about the DTLS_HT_HELLO_REQUEST with epoch 0 and a "DTLS_STATE_CONNECTED" client.

```
dtls_send_client_hello(ctx, peer, NULL, 0);
```

will send that CLIENT_HELLO in the current epoch (encrypted). That will be either dropped by the server, or, if processed, the server would have been able to send the HELLO_REQUEST in the current epoch (encrypted) as well.

**Comment by mrdeep1** [comment]
My understanding is that the server can send DTLS_HT_HELLO_REQUEST whenever it likes to cause the client to initiate a CLIENT_HELLO handshake sequence whenever it wants..  The function `dtls_renegotiate()` will do just this with I believe the HELLO_REQUEST encrypted under the appropriate epoch.  That should then trigger CLIENT_HELLO encrypted under the same epoch.
````
   When the server desires a rehandshake, it transitions from the
   FINISHED state to the PREPARING state to transmit the HelloRequest.
   When the client receives a HelloRequest, it transitions from FINISHED
   to PREPARING to transmit the ClientHello.
````

peer->handshake_params goes away when processing DTLS_HT_FINISHED which makes reading the code a bit confusing in some places.  However, I don't see when HELLO_REQUEST can be sent in epoch 0 (other than by a misbehaving entity).  `hs_attempt_with_existing_peer()` looks broken to me on that front.

**Comment by boaks** [comment]
> hs_attempt_with_existing_peer() looks broken to me on that front.

Agreed.

Using ALERTs or HELLO_REQUEST in epoch 0 seems to be used by some, which want to have faster fail-overs (which occurs  frequently on restarts and ip-address changes)  in trade of being vulnerable against spoofing (which occurs less frequently compared to ip-address changes). With that, I still don't want to use it, but if others want to do that aware of the risks, the it's up to them.
So for me the question seems to be, is there someone using this for epoch 0? And how is it then intended to work (assuming that other fixes afterwards have broken the initial implementation).

**Comment by boaks** [comment]
I tried to get some more information about the history of that from sourceforge .. in vain :-).

@obgm

Do you know something about using HELLO_REQUEST in epoch 0 (not compliant to RFC5246/RFC6347)?
Easiest to go will be removing it and wait for complains ;-).

**Comment by mrdeep1** [comment]
````
        if (msg_epoch == 0) {
          dtls_handshake_header_t * hs_header = DTLS_HANDSHAKE_HEADER(msg + DTLS_RH_LENGTH);
          if (hs_header->msg_type == DTLS_HT_CLIENT_HELLO ||
              hs_header->msg_type == DTLS_HT_HELLO_REQUEST) {
            return 1;
          }
````
looks suspiciously like a cut and paste from
````
    /* This is a ClientHello or Hello Request send when doing TLS renegotiation */
    if (hs_header->msg_type == DTLS_HT_CLIENT_HELLO ||
        hs_header->msg_type == DTLS_HT_HELLO_REQUEST) {
      return handle_handshake_msg(ctx, peer, session, role, state, data,
                                  data_length);
    } else {
````

**Comment by boaks** [comment]
I removed the rudiments of the HELLO_REQUEST in epoch 0.
If really someone complains, it may be introduced again, then with details on how that should be processed.

**Comment by boaks** [comment]
Fixed with PR #89
